docs: Design doc for Special Purpose WRBs BN (#2948)
hiero-ledger/hiero-block-node: docs/design/wrb-streaming/sp-wrb-bn-design.md

# Special-Purpose WRB Block Node — Design

## Table of Contents

1. [Purpose](#purpose)
2. [Goals](#goals)
3. [Terms](#terms)
4. [Entities](#entities)
5. [Design](#design)
6. [Diagram](#diagram)
7. [Configuration](#configuration)
8. [Metrics](#metrics)
9. [Exceptions](#exceptions)
10. [Acceptance Tests](#acceptance-tests)
11. [Open Questions](#open-questions)

## Purpose

Distribute historical and live **Wrapped Record Blocks (WRBs)** and **TSS** data from a WRB CLI
conversion server to Council-operated **Tier-1 Block Nodes**. A co-located **special-purpose Block

> Node** on acts as the initial source ("Tier 0"): the CLI pushes blocks into it through the normal

Reviewer: Minor type: Node acts on as.

ingestion path, and Tier-1 nodes pull blocks (backfill) and query TSS (status API) from it.
To serve *historical* WRBs, the Block Node must verify each block against the address book that
was in effect for that block, not just the current one.

This is intended as the durable long-term solution (historical address-book verification), not
throwaway migration code.

## Goals

- Push historical WRBs from the CLI's existing wrapped-block storage into the special-purpose BN,
  one block at a time, via the existing ingestion (publish) API — no disk-layout restructuring, no BN

>   restarts.

Reviewer 2: With Rocky's findings, I think this changes.

```suggestion
- Push historical WRBs from the CLI's existing wrapped-block storage into the special-purpose BN,
  via simple bulk load with the BN not running. The BN will detect the bulk data loaded on startup,
  and will accept blocks from that point forward on the Publish API.
```

Reviewer: +1 Agreed.

- Keep the special-purpose BN current during the transition by pushing each new WRB live as
  the CLI produces it (no 6-hour zip-batch floods).
- Verify any historical WRB by selecting the address book in effect for that block's number.
- Let Tier-1 nodes use the special-purpose BN as a backfill source for WRBs and TSS details.
- Reuse existing code where possible to reduce overhead and throwaway.

## Terms

<dl>
  <dt>WRB (Wrapped Record Block)</dt>
  <dd>A legacy Hedera record file wrapped into the Block Stream block format, carrying a
      <code>SignedRecordFileProof</code> verified against node RSA keys.</dd>
  <dt>TSS data</dt>
  <dd>Threshold-signature-scheme material (ledger id, roster, WRAPS verification key) the BN needs;
      produced by the CLI as <code>tss-bootstrap-roster.json</code>.</dd>
  <dt>Special-purpose BN ("Tier 0")</dt>
  <dd>A Block Node co-located with the CLI that serves as the initial WRB source and TSS peer
      for Tier-1 nodes, effectively replacing the consensus node for WRB distribution.</dd>

>   <dt>Address book roster</dt>

Reviewer 2: We overuse both Address Book and Roster in too many places already.
Perhaps rename slightly?

```suggestion
  <dt>Address Book History</dt>
```

  <dd>An ordered set of historical address books, each scoped to a block-number range, used to verify
      a WRB against the keys that were valid when its underlying record file was signed.</dd>
  <dt>CLI</dt>

>   <dd>The picocli tool <code>org.hiero.block.tools.BlockStreamTool</code> under

Reviewer 2: Picocli is already known as the Consensus Node piocli.
We need a little more text in the description here, I think.

```suggestion
  <dd>The picocli-derived tool <code>org.hiero.block.tools.BlockStreamTool</code> under
```

      <code>tools-and-tests/tools</code> (commands <code>blocks</code>, <code>days</code>,
      <code>mirror</code>, <code>networkCapacity</code>, …).</dd>
</dl>

## Entities

- **CLI (`tools-and-tests/tools`)** — already wraps record files into WRBs (`blocks wrap`,
  `days live-sequential`), maintains address-book history (`mirror generateAddressBook*`,
  `compareAddressBooks`), and can push blocks over gRPC (`networkCapacity` client). Extended here with

>   backfill-push and live-push.

Reviewer: This seems decent names. 

I assume backfill-push is the new command to copy over the archive.

The live-push is a config for `live-sequential`?

## Design

### 1. Push-based ingestion (CLI → special BN)

The CLI pushes blocks into the BN through the normal publish/ingestion endpoint so the BN performs its
usual validation and formatting; no on-disk restructuring and no restarts.

- **Backfill (one-time):** a CLI command reads the existing wrapped-block zips (10k blocks/zip,
  produced ~every 6 hours) and publishes them **one block at a time**, in order. Reuses the gRPC
  publish client already present for `networkCapacity` (and the simulator's
  `PublishStreamGrpcClient`). Resumable from the BN's latest block so a restart continues where it
  left off.
- **Live (continuous):** the CLI's live wrap pipeline (`days live-sequential` / `download-live2`) is
  extended to publish each WRB immediately as it is produced, bypassing the in-memory 6-hour batching
  for the distribution path. Reuses the same publish client.

One-block-at-a-time ingestion (vs. a 6-hour batch flood) is the safer, simpler operational model and
keeps the special-purpose BN continuously current.

### 2. Historical address-book roster (BN)

Today the BN verifies WRBs against a single, current address book (`RsaRosterBootstrapPlugin` loads one

> `NodeAddressBook`). To verify *historical* WRBs the BN holds a **roster of address books, each scoped

Reviewer 2: To limit confusion, we may want to avoid calling it a roster.

```suggestion
`NodeAddressBook`). To verify *historical* WRBs the BN holds a **history of address books, each scoped
```

to a `[startBlock, endBlock]` range**:

- **Model + store:** an ordered, range-keyed collection with O(log n) lookup by block number; extends
  `RsaRosterBootstrapPlugin` / `RsaRosterBootstrapConfig`.
- **Bootstrap:** load the CLI's historical address-book history (the ~20 JSON files the CLI already
  maintains) converted into the roster format. The conversion is an operator script/CLI step (see
  §Configuration and the tickets).
- **Maintenance:** the BN queries Mirror Node for address-book changes from genesis forward, translates
  the change's consensus time to a block number, and appends a new roster entry — extending the RSA
  plugin's existing Mirror-Node query path.

### 3. Historical WRB verification by block number

WRB verification (`ExtendedMerkleTreeSession`, today a single `rsaKeyByNodeId` map) selects the address
book whose range contains the block being verified, and validates the `SignedRecordFileProof` against

> that era's node RSA keys. A block with no matching roster entry fails verification with a clear cause.

Reviewer 2: ```suggestion
that era's node RSA keys. A block with no available address book fails verification with a clear cause.
```

### 4. TSS distribution (no new code)

Tier-1 retrieval already exists via `RosterBootstrapTssPlugin.queryPeerTssData()` against the
special-purpose BN's status API. The only operational step is moving the CLI-produced
`/mnt/wrb-operations/wrappedBlocks/tss-bootstrap-roster.json` to the BN's
`/opt/hiero/block-node/node/` — a documented operator/script action, **not** an engineering change.

> Multiple processes must not write the same location; use a copy/move, not a shared path.

Reviewer 2: Or better, just stop the block node before copying the file, then restart after.

  already-stored blocks).

## Acceptance Tests

- Backfill pushes the full historical WRB set into the special-purpose BN, one block at a time, and is
  resumable after interruption.
- Live push keeps the special-purpose BN current as the CLI produces new WRBs (no 6-hour gaps).
- A historical WRB from an earlier address-book era verifies successfully against the correct era's
  roster entry; a block outside any roster range fails with the documented cause.
- A Tier-1 BN backfills blocks from the special-purpose BN and verifies them.
- After the operator moves `tss-bootstrap-roster.json` into the BN config dir, a Tier-1 BN's
  `queryPeerTssData()` retrieves TSS data successfully.

## Open Questions

- **Address-book file delivery:** ship the roster as a code resource vs. require operator loading
  (leaning toward a documented script/tool that converts the CLI's address-book history).
- **Backfill-source API plugin:** whether the special-purpose BN should run a simplified API plugin
  set to minimize failure points (discussed, undecided).
- **Consensus-time → block-number translation:** authoritative source/precision for mapping address-book

>   change times to block ranges (CLI block-times data vs. Mirror Node).

Reviewer 2: For BN this really needs to be block _number_ as getting the consensus time requires a lot more data parsing than we would prefer.
